vlog4j-parser/src/test/java/org/semanticweb/vlog4j/syntax/parser/EntityTest.java

```diff
@@ -0,0 +1,124 @@
+import org.semanticweb.vlog4j.parser.ParsingException;
+import org.semanticweb.vlog4j.parser.RuleParser;
+
+public class EntityTest {
+	final Variable x = Expressions.makeUniversalVariable("X");
+	final Variable y = Expressions.makeUniversalVariable("Y");
+	final Variable z = Expressions.makeExistentialVariable("Z");
+	final Variable y2 = Expressions.makeUniversalVariable("Y");
+	final Constant d = Expressions.makeAbstractConstant("d");
+	final Constant c = Expressions.makeAbstractConstant("c");
+	final AbstractConstantImpl f = new AbstractConstantImpl("f");
+	final LanguageStringConstantImpl s = new LanguageStringConstantImpl("Test", "en");
+	final DatatypeConstantImpl data = new DatatypeConstantImpl("data", "http://example.org/mystring");
+	final PositiveLiteral atom1 = Expressions.makePositiveLiteral("p", x, c);
+	final PositiveLiteral atom2 = Expressions.makePositiveLiteral("p", x, y);
+	final PositiveLiteral headAtom1 = Expressions.makePositiveLiteral("q", x, z);
+	final PositiveLiteral positiveLiteral1 = Expressions.makePositiveLiteral("p", x, c);
+	final PositiveLiteral positiveLiteral2 = Expressions.makePositiveLiteral("p", y2, x);
+	final PositiveLiteral positiveLiteral3 = Expressions.makePositiveLiteral("q", x, d);
+	final NegativeLiteral NegativeLiteral = Expressions.makeNegativeLiteral("r", x, d);
+	final PositiveLiteral PositiveLiteral4 = Expressions.makePositiveLiteral("s", c, s);
+	final Predicate p = Expressions.makePredicate("p", 2);
+	final Fact f1 = Expressions.makeFact(p, Arrays.asList(f, s));
+	final Fact f2 = Expressions.makeFact("p", Arrays.asList(data, d));
+	final List<Literal> LiteralList = Arrays.asList(positiveLiteral1, positiveLiteral2, positiveLiteral3,
+			NegativeLiteral, PositiveLiteral4);
+	final Conjunction<Literal> bodyLiterals = Expressions.makeConjunction(atom1, atom2);
+	final Conjunction<PositiveLiteral> headPositiveLiterals = Expressions.makePositiveConjunction(headAtom1);
+	final Conjunction<Literal> bodyConjunction = new ConjunctionImpl<>(LiteralList);
+	final Rule rule1 = new RuleImpl(headPositiveLiterals, bodyLiterals);
+	final Rule rule2 = new RuleImpl(headPositiveLiterals, bodyConjunction);
+
+	@Test
+	public void factToStringRoundTripTest() throws ParsingException {
+		assertEquals(RuleParser.parseFact(f1.toString()), RuleParser.parseFact("p(f, \"Test\"@en)."));
+		assertEquals(RuleParser.parseFact(f2.toString()),
+				RuleParser.parseFact("p(\"data\"^^<http://example.org/mystring>, d)."));
+	}
+
+	@Test
+	public void ruleToStringRoundTripTest() throws ParsingException {
+		assertEquals(RuleParser.parseRule(rule1.toString()), RuleParser.parseRule("q(?X, !Z) :- p(?X, c), p(?X, ?Y)."));
+		assertEquals(RuleParser.parseRule(rule2.toString()),
+				RuleParser.parseRule("q(?X, !Z) :- p(?X, c), p(?Y, ?X), q(?X, d), ~r(?X, d), s(c, \"Test\"@en)."));
+	}
+
+	@Test
+	public void dataSourceDeclarationToStringParsingTest() throws ParsingException, IOException {
+		KnowledgeBase kb = new KnowledgeBase();
```

Comment:
I would simplify the unit tests, they are quite difficult to follow now:
- test only one object per test method
- compare a string with the toString on the object that results from parsing that string

Reply:
The only way to test this is through a rule, shouldn't be easier to create a rule with all different types and test them, instead of just creating new rule for every type as it is pointless to do so ?

Reply:
You are right. Then maybe have the fields as local variables, if they are not used in other methods.